### DISCUSSION TITLE
BackgroundTasks can't work with middleware

### DISCUSSION THREAD

**Discussion by skk294** [fastapi/fastapi#8815] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import asyncio
from fastapi import Request, BackgroundTasks, FastAPI
from fastapi.middleware.cors import CORSMiddleware

app = FastAPI()

origins = ["*"]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


async def test_sleep():
    print('test')
    await asyncio.sleep(9999)


@app.post("/test")
async def test(background_tasks: BackgroundTasks):
    background_tasks.add_task(test_sleep)
    return 'test'


@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    response = await call_next(request)
    return response
```

### Description

When use middleware, the second request to 'http://127.0.0.1:8000/test' will be pending. If I remove middleware, everything works fine.
I know it's the middleware problem, but I don't known how to fix it.

### Operating System

Linux

### Operating System Details

Ubuntu 20.04

### FastAPI Version

0.66.1

### Python Version

3.9.5

**Comment by STeveShary** [comment]
Which middleware do you remove to get it to work?  The Cors Middleware or your http one?  (or both?)

**Comment by skk294** [comment]
> Which middleware do you remove to get it to work? The Cors Middleware or your http one? (or both?)

http

**Comment by STeveShary** [comment]
@skk294 I spent some time to try and reproduce this issue.  Everything is working as expected.  This is what I have tried:

I put your test application into `test.py` and added a print statement to verify the middleware is running.  Below is the code I have in `test.py`
```python3
import asyncio
from fastapi import Request, BackgroundTasks, FastAPI
from fastapi.middleware.cors import CORSMiddleware

app = FastAPI()

origins = ["*"]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


async def test_sleep():
    print('test')
    await asyncio.sleep(9999)


@app.post("/test")
async def test(background_tasks: BackgroundTasks):
    background_tasks.add_task(test_sleep)
    return 'test'


@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    response = await call_next(request)
    print("Finished middleware.")
    return response

```
- I then installed `uvicorn` to serve my test application. 
- After that, I started the appplication with the command `uvicorn test:app`
- To test the API, I then ran the command: `curl -X POST http://localhost:8000/test`

Everything worked well and I saw the following responses in the logs:
```
INFO:     Started server process [80332]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
test
Finished middleware.
INFO:     127.0.0.1:54165 - "POST /test HTTP/1.1" 200 OK
test
Finished middleware.
INFO:     127.0.0.1:54167 - "POST /test HTTP/1.1" 200 OK
test
Finished middleware.
INFO:     127.0.0.1:54169 - "POST /test HTTP/1.1" 200 OK
test
Finished middleware.
INFO:     127.0.0.1:54171 - "POST /test HTTP/1.1" 200 OK
test
Finished middleware.
INFO:     127.0.0.1:54173 - "POST /test HTTP/1.1" 200 OK
test
Finished middleware.
INFO:     127.0.0.1:54175 - "POST /test HTTP/1.1" 200 OK
test
Finished middleware.
INFO:     127.0.0.1:54177 - "POST /test HTTP/1.1" 200 OK
test
Finished middleware.
INFO:     127.0.0.1:54179 - "POST /test HTTP/1.1" 200 OK
test
Finished middleware.
INFO:     127.0.0.1:54181 - "POST /test HTTP/1.1" 200 OK
```

Can you try the test described above and let me know what you get?

**Comment by skk294** [comment]
> @skk294 I spent some time to try and reproduce this issue. Everything is working as expected. This is what I have tried:
> 
> I put your test application into `test.py` and added a print statement to verify the middleware is running. Below is the code I have in `test.py`
> 
> ```python
> import asyncio
> from fastapi import Request, BackgroundTasks, FastAPI
> from fastapi.middleware.cors import CORSMiddleware
> 
> app = FastAPI()
> 
> origins = ["*"]
> 
> app.add_middleware(
>     CORSMiddleware,
>     allow_origins=origins,
>     allow_credentials=True,
>     allow_methods=["*"],
>     allow_headers=["*"],
> )
> 
> 
> async def test_sleep():
>     print('test')
>     await asyncio.sleep(9999)
> 
> 
> @app.post("/test")
> async def test(background_tasks: BackgroundTasks):
>     background_tasks.add_task(test_sleep)
>     return 'test'
> 
> 
> @app.middleware("http")
> async def add_process_time_header(request: Request, call_next):
>     response = await call_next(request)
>     print("Finished middleware.")
>     return response
> ```
> 
> * I then installed `uvicorn` to serve my test application.
> * After that, I started the appplication with the command `uvicorn test:app`
> * To test the API, I then ran the command: `curl -X POST http://localhost:8000/test`
> 
> Everything worked well and I saw the following responses in the logs:
> 
> ```
> INFO:     Started server process [80332]
> INFO:     Waiting for application startup.
> INFO:     Application startup complete.
> INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
> test
> Finished middleware.
> INFO:     127.0.0.1:54165 - "POST /test HTTP/1.1" 200 OK
> test
> Finished middleware.
> INFO:     127.0.0.1:54167 - "POST /test HTTP/1.1" 200 OK
> test
> Finished middleware.
> INFO:     127.0.0.1:54169 - "POST /test HTTP/1.1" 200 OK
> test
> Finished middleware.
> INFO:     127.0.0.1:54171 - "POST /test HTTP/1.1" 200 OK
> test
> Finished middleware.
> INFO:     127.0.0.1:54173 - "POST /test HTTP/1.1" 200 OK
> test
> Finished middleware.
> INFO:     127.0.0.1:54175 - "POST /test HTTP/1.1" 200 OK
> test
> Finished middleware.
> INFO:     127.0.0.1:54177 - "POST /test HTTP/1.1" 200 OK
> test
> Finished middleware.
> INFO:     127.0.0.1:54179 - "POST /test HTTP/1.1" 200 OK
> test
> Finished middleware.
> INFO:     127.0.0.1:54181 - "POST /test HTTP/1.1" 200 OK
> ```
> 
> Can you try the test described above and let me know what you get?

![Screenshot from 2021-09-11 09-24-24](https://user-images.githubusercontent.com/65146378/132932054-fccafd90-1f12-4f14-bced-0981be9c0243.png)

![Screenshot from 2021-09-11 09-25-33](https://user-images.githubusercontent.com/65146378/132932126-72f8aff7-1984-41ca-9042-617bbed61569.png)

```
(fastapi_test) skk@skk:~$ pip list
Package           Version
----------------- --------
asgiref           3.4.1
click             8.0.1
fastapi           0.68.1
h11               0.12.0
pip               21.1.1
pydantic          1.8.2
setuptools        56.0.0
starlette         0.14.2
typing-extensions 3.10.0.2
uvicorn           0.15.0
```

To test it, I created a new venv, so the fastapi version is 0.68.1. And I also test it in previous venv whose fastapi version is 0.66.1 uvicorn version 0.14.0, same problem.
But use `curl -X POST http://localhost:8000/test` works fine.
I got this problem because I was writing a website using fastapi, the user add backgrounp task via POST method, and it will somehow block the next requested url, but the third and the following request won't be block.

**Comment by havardthom** [comment]
This is a known starlette issue https://github.com/encode/starlette/issues/919

**Comment by ghost** [comment]
I'm seeing a probably related issue but with different symptoms than skk294 or the ones described in the starlette issue. Here is my code:
```python
import asyncio
from fastapi import FastAPI, BackgroundTasks, Request

app = FastAPI()

@app.get("/")
def endpoint(bg: BackgroundTasks):
	bg.add_task(f)

async def f():
	print('starting background task')
	await asyncio.sleep(1)
	print('finished background task')

@app.middleware("http")
async def db_session_middleware(request: Request, call_next):
    print('starting middleware')
    response = await call_next(request)
    print('finishing middleware')
    return response
```

Python 3.10.4, Fastapi 0.78.0, Uvicorn 0.17.6

The symptom is that the background task hangs as soon as it encounters `await asyncio.sleep`. Future requests will go the same way: I see "finishing middleware", and "starting background task", but never "finished background task". If I remove the middleware, it works.

Interestingly this only happens with asyncio: if I replace asyncio.sleep with time.sleep (and remove the await), it works even with the middleware.

**Comment by aguckenber-chwy** [comment]
Not merged yet but if anyone reads this far and asks

> Gosh, I wish these raw middlewares were better documented!

Go checkout this PR:
https://github.com/encode/starlette/pull/1656/files

**Comment by OmriTreidel** [comment]
@yujiri8 I'm having the same issue with Python 3.8.6, fastapi==0.85.1, starlette==0.20.4 and uvicorn==0.19.0. 

Did you find a way around it?

**Comment by Kludex** [comment, marked as answer]
This is already solved on the latest FastAPI version.
